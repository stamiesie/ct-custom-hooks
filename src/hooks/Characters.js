import { useEffect, useState } from 'react';
import { fetchCharacters } from '../services/ApiCall';

export const useAllCharacters = () => {
  const [loading, setLoading] = useState(true);
  const [characters, setCharacters] = useState([]);

  useEffect(() => {
    fetchCharacters()
      .then((characters) => setCharacters(characters))
      .finally(() => setLoading(false));
  }, []);

  return { loading, characters };
};

