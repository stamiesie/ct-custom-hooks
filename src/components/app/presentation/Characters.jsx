import React from 'react';
import CharacterList from './CharacterList';
import { useAllCharacters } from '../../../hooks/Characters';
import Loading from '../../../loading/Loading';

const Characters = () => {
  const { loading, characters } = useAllCharacters();

  if(loading) return <Loading />;

  return <CharacterList characters={characters}/>;

};

export default Characters;
