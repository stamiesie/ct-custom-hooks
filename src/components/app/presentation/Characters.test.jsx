import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import mockCharacterData from '../../../../fixtures/AllCharacters.json';
import Characters from './Characters';
import { MemoryRouter } from 'react-router-dom';

const server = setupServer(
  rest.get('https://hey-arnold-api.herokuapp.com/api/v1/characters', 
    (req, res, ctx) => {
      return res(ctx.json(mockCharacterData));
    })
);

describe('All Characters custom hook', () => {
  beforeAll(() => server.listen());
  afterAll(() => server.close());

  it('renders a list of characters to the page', async () => {
    render(
      <MemoryRouter>
        <Characters />
      </MemoryRouter>
    );

    screen.getByText('Loading...');

    const ul = await screen.findByRole('list', { name: 'characters' });
    expect(ul).not.toBeEmptyDOMElement();
    expect(ul).toMatchSnapshot();

    return waitFor(() => {
      screen.getByText('Nancy', { exact: false });
    });
  });
});
