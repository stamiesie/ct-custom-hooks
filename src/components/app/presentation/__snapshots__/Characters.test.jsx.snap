// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Characters custom hook renders a list of characters to the page 1`] = `
<ul
  aria-label="characters"
>
  <li>
    <a
      href="/5da237699734fdcb7bef8f5c"
    >
      <div>
        <p>
          Nancy
        </p>
        <img
          alt="Nancy"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/6/62/Nancy.jpg/revision/latest/scale-to-width-down/310?cb=20110423055923"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f59"
    >
      <div>
        <p>
          Grandpa's grandfather
        </p>
        <img
          alt="Grandpa's grandfather"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/f/f7/Grandpa%27s_grandfather.png/revision/latest/scale-to-width-down/310?cb=20190228081015"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f52"
    >
      <div>
        <p>
          Grandpa Phil
        </p>
        <img
          alt="Grandpa Phil"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/7/79/GrandpaPhil.png/revision/latest/scale-to-width-down/300?cb=20160924030436"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f58"
    >
      <div>
        <p>
          Grandpa's father
        </p>
        <img
          alt="Grandpa's father"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/e/e0/Grandpa%27s_father.png/revision/latest/scale-to-width-down/310?cb=20190228081112"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f70"
    >
      <div>
        <p>
          Phoebe Heyerdahl
        </p>
        <img
          alt="Phoebe Heyerdahl"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/9/93/Phoebe_%28classic%29.png/revision/latest/scale-to-width-down/247?cb=20190816022206"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f7e"
    >
      <div>
        <p>
          Eugene Horowitz
        </p>
        <img
          alt="Eugene Horowitz"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/9/99/EAFDB121-EC13-4F12-849E-37037A95E3D6.png/revision/latest/scale-to-width-down/270?cb=20190816022130"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f82"
    >
      <div>
        <p>
          Ray
        </p>
        <img
          alt="Ray"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/3/34/Ray.png/revision/latest/scale-to-width-down/310?cb=20181006091830"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f8b"
    >
      <div>
        <p>
          Shelley
        </p>
        <img
          alt="Shelley"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/c/cb/Shelley.jpg/revision/latest/scale-to-width-down/310?cb=20120527025256"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f76"
    >
      <div>
        <p>
          Lila Sawyer
        </p>
        <img
          alt="Lila Sawyer"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/b/b0/Lila_Sawyer.png/revision/latest/scale-to-width-down/310?cb=20171212053404"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f7c"
    >
      <div>
        <p>
          Uncle Stinky
        </p>
        <img
          alt="Uncle Stinky"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/8/89/Uncle_Stinky.jpg/revision/latest/scale-to-width-down/310?cb=20120924204406"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f55"
    >
      <div>
        <p>
          Arnie
        </p>
        <img
          alt="Arnie"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/4/42/Arnie.jpg/revision/latest/scale-to-width-down/310?cb=20110109195825"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f66"
    >
      <div>
        <p>
          Olga Pataki
        </p>
        <img
          alt="Olga Pataki"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/5/5e/Olga_Pataki.png/revision/latest/scale-to-width-down/310?cb=20180124053913"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f6b"
    >
      <div>
        <p>
          Timberly Johanssen
        </p>
        <img
          alt="Timberly Johanssen"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/e/e8/Timberly.png/revision/latest/scale-to-width-down/310?cb=20171210045406"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f6d"
    >
      <div>
        <p>
          Jerry Berman
        </p>
        <img
          alt="Jerry Berman"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/2/2f/Jerry_Berman.png/revision/latest/scale-to-width-down/310?cb=20171219061306"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f73"
    >
      <div>
        <p>
          Rhonda Wellington Lloyd
        </p>
        <img
          alt="Rhonda Wellington Lloyd"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/b/b4/C1BBB448-FE38-4F7B-98FA-2137F277F9C5.png/revision/latest/scale-to-width-down/213?cb=20190816022649"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f78"
    >
      <div>
        <p>
          Stinky Peterson
        </p>
        <img
          alt="Stinky Peterson"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/2/24/Stinky_Peterson.png/revision/latest/scale-to-width-down/310?cb=20160516072820"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f71"
    >
      <div>
        <p>
          Kyo Heyerdahl
        </p>
        <img
          alt="Kyo Heyerdahl"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/c/cf/Kyo_Heyerdahl.png/revision/latest/scale-to-width-down/310?cb=20190310221807"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f8e"
    >
      <div>
        <p>
          Iggy
        </p>
        <img
          alt="Iggy"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/0/0e/Iggy.png/revision/latest/scale-to-width-down/310?cb=20190210045957"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f77"
    >
      <div>
        <p>
          Mr. Sawyer
        </p>
        <img
          alt="Mr. Sawyer"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/4/45/Mr._Sawyer.png/revision/latest/scale-to-width-down/310?cb=20171219055204"
        />
      </div>
    </a>
  </li>
  <li>
    <a
      href="/5da237699734fdcb7bef8f84"
    >
      <div>
        <p>
          Thaddeus "Curly" Gammelthorpe
        </p>
        <img
          alt="Thaddeus \\"Curly\\" Gammelthorpe"
          src="https://vignette.wikia.nocookie.net/heyarnold/images/b/ba/Classic_Curly_promo.png/revision/latest/scale-to-width-down/246?cb=20181010011614"
        />
      </div>
    </a>
  </li>
</ul>
`;
